File: apple/Sources/BidiLens/BidiMath.swift
import Foundation

extension BidiAnalyzer {
    // Forward-only math heuristic, matching the core dollar-boundary policy.
    static func mathRanges(_ text: String) -> [TechnicalTokenRange] {
        guard text.contains("$") || text.contains("\\(") else { return [] }
        let units = Array(text.utf16)
        let delimiters: [[UInt16]] = [[36], [36, 36], [92, 41]]
        var scanned = [-1, -1, -1]
        var ranges: [TechnicalTokenRange] = []
        var i = 0
        func at(_ delimiter: [UInt16], _ index: Int) -> Bool {
            index + delimiter.count <= units.count
                && units[index..<(index + delimiter.count)].elementsEqual(delimiter)
        }
        func space(_ unit: UInt16) -> Bool {
            // Pin ECMAScript whitespace instead of a host-specific predicate.
            switch unit {
            case 0x09...0x0d, 0x2000...0x200a, 0x20, 0xa0, 0x1680, 0x2028, 0x2029, 0x202f, 0x205f, 0x3000, 0xfeff:
                return true
            default: return false
            }
        }
        while i < units.count {
            let paren = at([92, 40], i)
            if units[i] == 92 && !paren { i += 2; continue }
            let kind = units[i] == 36 ? (at([36, 36], i) ? 1 : 0) : (paren ? 2 : -1)
            if kind < 0 || i < scanned[kind] { i += 1; continue }
            let delimiter = delimiters[kind]
            if kind == 0 && (i + 1 == units.count || space(units[i + 1])) { i += 1; continue }
            var end = i + delimiter.count
            while end < units.count && units[end] != 13 && units[end] != 10 && !at(delimiter, end) {
                end += units[end] == 92 && end + 1 < units.count && units[end + 1] != 13 && units[end + 1] != 10 ? 2 : 1
            }
            if at(delimiter, end) && (kind != 0 || end > i + 1) {
                let next: UInt16 = end + 1 < units.count ? units[end + 1] : 0
                if kind == 0 && (space(units[end - 1]) || (48...57).contains(next)
                    || (0x660...0x669).contains(next) || (0x6f0...0x6f9).contains(next)) {
                    i = end
                    continue
                }
                let range = i..<(end + delimiter.count)
                ranges.append(TechnicalTokenRange(
                    text: UnicodeClassifier.substring(text, utf16Range: range),
                    utf16Range: range, kind: .math
                ))
                i = range.upperBound
            } else {
                scanned[kind] = end
                i += 1
            }
        }
        return ranges
    }
}

File: windows/src/BidiLens.Core/BidiMath.cs
namespace BidiLens;

public static partial class BidiAnalyzer
{
    // Pin ECMAScript whitespace for parity; host char.IsWhiteSpace differs.
    private static bool IsMathWhitespace(char value) => value is >= '\u0009' and <= '\u000d'
        or >= '\u2000' and <= '\u200a' or ' ' or '\u00a0' or '\u1680' or '\u2028'
        or '\u2029' or '\u202f' or '\u205f' or '\u3000' or '\ufeff';
    // Forward-only math heuristic, matching the core dollar-boundary policy.
    private static void AddMathRanges(string text, List<TechnicalTokenRange> ranges)
    {
        var scanned = new Dictionary<string, int> { ["$"] = -1, ["$$"] = -1, [@"\)"] = -1 };
        var i = 0;
        bool At(string delimiter, int index) => text.AsSpan(index).StartsWith(delimiter, StringComparison.Ordinal);
        while (i < text.Length)
        {
            var paren = At(@"\(", i);
            if (text[i] == '\\' && !paren) { i += 2; continue; }
            var delimiter = text[i] == '$' ? (At("$$", i) ? "$$" : "$") : paren ? @"\)" : "";
            if (delimiter.Length == 0 || i < scanned[delimiter]) { i++; continue; }
            if (delimiter == "$" && (i + 1 == text.Length || IsMathWhitespace(text[i + 1]))) { i++; continue; }
            var end = i + (paren ? 2 : delimiter.Length);
            while (end < text.Length && text[end] is not ('\r' or '\n') && !At(delimiter, end))
                end += text[end] == '\\' && end + 1 < text.Length && text[end + 1] is not ('\r' or '\n') ? 2 : 1;
            if (At(delimiter, end) && (delimiter != "$" || end > i + 1))
            {
                var next = end + 1 < text.Length ? text[end + 1] : '\0';
                if (delimiter == "$" && (IsMathWhitespace(text[end - 1])
                    || next is >= '0' and <= '9' or >= '\u0660' and <= '\u0669' or >= '\u06f0' and <= '\u06f9'))
                {
                    i = end;
                    continue;
                }
                ranges.Add(new(text[i..(end + delimiter.Length)], i, end + delimiter.Length, TechnicalTokenKind.Math));
                i = end + delimiter.Length;
            }
            else
            {
                scanned[delimiter] = end;
                i++;
            }
        }
    }
}
